107/src/main/java/org/ehcache/jsr107/EhcacheCachingProvider.java

```diff
@@ -108,14 +113,53 @@ public CacheManager getCacheManager(URI uri, Configuration config, Properties pr
     return getCacheManager(new ConfigSupplier(uri, config), properties);
   }
 
-  Eh107CacheManager getCacheManager(ConfigSupplier configSupplier, Properties properties) {
+  CacheManager getCacheManager(ConfigSupplier configSupplier, Properties properties) {
+    ClassLoader classLoader = configSupplier.getClassLoader();
+    URI uri = configSupplier.getUri();
+
+    return getCacheManager(uri, classLoader, () -> createCacheManager(uri, configSupplier.getConfiguration(), properties));
+  }
+
+  private Eh107CacheManager createCacheManager(URI uri, Configuration config, Properties properties) {
+    Eh107CacheLoaderWriterProvider cacheLoaderWriterFactory = new Eh107CacheLoaderWriterProvider();
+
+    Collection<ServiceCreationConfiguration<?>> serviceCreationConfigurations = config.getServiceCreationConfigurations();
+
+    Jsr107Service jsr107Service = new DefaultJsr107Service(ServiceUtils.findSingletonAmongst(Jsr107Configuration.class, serviceCreationConfigurations));
+
+    Collection<Service> services = new ArrayList<>(4);
+    services.add(cacheLoaderWriterFactory);
+    services.add(jsr107Service);
+
+    if (ServiceUtils.findSingletonAmongst(DefaultSerializationProviderConfiguration.class, serviceCreationConfigurations) == null) {
+      services.add(new DefaultJsr107SerializationProvider());
+    }
+
+    org.ehcache.CacheManager ehcacheManager = new EhcacheManager(config, services, !jsr107Service.jsr107CompliantAtomics());
+    ehcacheManager.init();
+
+    return new Eh107CacheManager(this, ehcacheManager, jsr107Service, properties, config.getClassLoader(), uri,
+            new ConfigurationMerger(config, jsr107Service, cacheLoaderWriterFactory));
+  }
+
+  /**
+   * Enables to create a JSR-107 {@link CacheManager} based on the provided Ehcache {@link CacheManagerBuilder}.
+   *
+   * @param uri the URI identifying this cache manager
+   * @param classLoader class loader where the cache manager will be registered
+   * @param builder the builder to create the underlying EhcacheManager
+   *
+   * @return a cache manager
+   */
+  public CacheManager getCacheManager(URI uri, ClassLoader classLoader, CacheManagerBuilder<org.ehcache.CacheManager> builder) {
+    return getCacheManager(uri, classLoader, () -> createCacheManager(uri, classLoader, builder));
+  }
+
+  private CacheManager getCacheManager(URI uri, ClassLoader classLoader, Supplier<Eh107CacheManager> cacheManagerCreator) {
```

Comment:
It feel like this Supplier should really be a function, possibly taking the `URI` and `ClassLoader` as arguments.

Reply:
I tried but it makes things redundant. With lines like `return getCacheManager(uri, classLoader, (u, c) -> createCacheManager(u, c, builder));`. But I see what you mean. It removes the closure for these two params by the caller. But I still have a closure. Still prefer me to change it?